ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -150,6 +150,9 @@ static size_t tsrm_tls_offset;
 
 #define IS_32BIT(addr) (((uintptr_t)(addr)) <= 0x7fffffff)
 
+/* Call range is before or after 2GB */
+#define IS_NEAR(end_addr, func) ((((uintptr_t)(end_addr)^(uintptr_t)(func)) >> 33) == 0)
+
```

Comment:
I didn't understand this condition. Why shift to 33 bits?
I think it also should take care about ``dasm_buf``
I would propose:

```c
#define MAY_USE_32BIT_ADDR(addr) \
	(IS_SIGNED_32BIT((char*)(addr) - (char*)dasm_buf) && \
	IS_SIGNED_32BIT((char*)(addr) - (char*)dasm_end))
```

Reply:
@dstogov Thanks for the proposal and we will use it. And we did make a big mistake here for right shifting 33 bits and we actually should right shift 31 bits, because what we are trying to check is that **the highest 33 bits must all be zeros after XOR operation**.

Reply:
@dstogov  Thanks for your proposal.  How about this implementation? Bit operation is a little bit faster, though with less code readability.
```
#define IS_NEAR(end_addr, func) ((((uintptr_t)(end_addr)^(uintptr_t)(func)) >> 31) == 0)

||		if (IS_NEAR(dasm_buf, func) && IS_NEAR(dasm_end, func)) {
|			call qword &func
||		}
```

Reply:
Your implementation guarantees that the top 33 bits are the same, but this check is not accurate enough.
```
0x80000000 - 0x7fff0000 =  0x10000 (valid 32-bit offset)
(0x80000000 ^ 0x7fff0000) >> 31 = 1 (your implementation thinks it's invalid)
```
It's possible to create a faster check, but this is not very critical operation.

I wouldn't use name "NEAR" to avoid mess with far/near addresses in 16-bit x86 terminology.

Reply:
My fault and thanks a lot. I updated it.